client_lib/pulp/client/launcher.py

```diff
@@ -63,29 +68,46 @@ def main(config, exception_handler_class=ExceptionHandler):
     prompt = _create_prompt(config)
     exception_handler = exception_handler_class(prompt, config)
 
-    # REST Bindings
-    username = options.username
-    password = options.password
-
-    if not username and not password:
-        # Try to get username/password from config if not explicitly set. username and password are
-        # not included by default so we need to catch KeyError Exceptions.
+    for auth_type in ['Kerberos', 'Basic']:
+        if auth_type is 'Kerberos':
+            if krb is not None and has_kerberos_ticket():
+                __, krb_context = krb.authGSSClientInit("HTTP@%s" % config['server']['host'])
+                krb.authGSSClientStep(krb_context, "")
+                negotiate_details = krb.authGSSClientResponse(krb_context)
+                logger.info("KERBEROS selected - %s" % krbV.default_context().default_ccache().principal())
+                username = krbV.default_context().default_ccache().principal()
+                password = None
```

Comment:
It seems like this pattern is repeating in several of these new code blocks. Perhaps it would be worth refactoring this into a function that they can all share together.